[stdlib] Create Box[] type as a new smart pointer

szbergeron opened the pull request:
This introduces a new Boxed[] type as an alternative to UnsafePointer[] to explicitly heap allocate large or slow to move/slow to copy types.

szbergeron committed:
Start work on boxed

Signed-off-by: Sawyer Bergeron <sawyerbergeron@gmail.com>

szbergeron committed:
Finish up Boxed[]

Signed-off-by: Sawyer Bergeron <sawyerbergeron@gmail.com>

szbergeron committed:
Fix up docstrs

Signed-off-by: Sawyer Bergeron <sawyerbergeron@gmail.com>

martinvuyk reviewed (commented):
Hi, thanks for the contribution. 

I'm not sure what the stdlib team's plans are with Reference and ownership of the underlying data, so this needs their input.

Personally, I like Rust's name `Box` more than `Boxed`. But the name I think does not do justice to what it really is: an `OwnedReference`

szbergeron committed:
Integrate comments from PR

Signed-off-by: Sawyer Bergeron <sawyerbergeron@gmail.com>

szbergeron committed:
Formatting

Signed-off-by: Sawyer Bergeron <sawyerbergeron@gmail.com>

szbergeron committed:
Fix docstr

Signed-off-by: Sawyer Bergeron <sawyerbergeron@gmail.com>

szbergeron committed:
Add period

Signed-off-by: Sawyer Bergeron <sawyerbergeron@gmail.com>

szbergeron committed:
Format boxed test

Signed-off-by: Sawyer Bergeron <sawyerbergeron@gmail.com>

szbergeron committed:
Merge branch 'nightly' into boxed_01

JoeLoser reviewed (commented):
Hey there, thanks for the contribution!  This is great to see coming together.  Internally, the team discussed this pull request at length in a full API design review this afternoon.  Here is some summarized feedback along with specific inline comments.

1. We prefer the name `Box` over `Boxed`.  It's shorter and the `ed` suffix doesn't provide a lot of additional value.  This is especially important for such common vocabulary types like `Box`.
2. Remove the `address_space` parameter on `Box` and just fix it to generic address space for the `UnsafePointer`.  Rationale: it's meaningless to provide this at the moment and only provides _confusion_ to users I'd argue.  In the future, once we design custom allocators, we can focus on this (perhaps address space will be a property on custom allocators anyways).  In short, let's lazily evaluate and add it later.
3. Let's have the invariant that `Box[T]` always has an initialized instance of `T`.  This will help us avoid the runtime branch in the destructor.  I'll comment inline on using `lit.ownership.mark_destroyed` in the consuming `take` method so we can avoid the destructor being run later (which is the need for the `if` check).

szbergeron committed:
Initial pass: rename Boxed -> Box, remove addrspace param

Signed-off-by: Sawyer Bergeron <sawyerbergeron@gmail.com>

szbergeron committed:
Integrate suggestions from PR

Signed-off-by: Sawyer Bergeron <sawyerbergeron@gmail.com>

szbergeron committed:
Fix some nits (and remove very trivially unreachable assert)

Signed-off-by: Sawyer Bergeron <sawyerbergeron@gmail.com>

szbergeron committed:
Lift ObservableDel, repoint test_arc at shared type

Signed-off-by: Sawyer Bergeron <sawyerbergeron@gmail.com>

szbergeron committed:
Repoint test_variant at shared ObservableDel type

Signed-off-by: Sawyer Bergeron <sawyerbergeron@gmail.com>

szbergeron committed:
Merge branch 'nightly' into boxed_01

szbergeron committed:
Merge branch 'nightly' into boxed_01

szbergeron committed:
Lifetime of refs handed out by Box[] should correspond to lt of `self`

Signed-off-by: Sawyer Bergeron <sawyerbergeron@gmail.com>

szbergeron committed:
Add Box[] to changelog (unreleased)

Signed-off-by: Sawyer Bergeron <sawyerbergeron@gmail.com>

szbergeron committed:
Remove misunderstood debug-info arg from box test

Signed-off-by: Sawyer Bergeron <sawyerbergeron@gmail.com>

szbergeron committed:
Add copying constructor to Box[]

Uses a dummy param to lower the specificity of
this constructor below the moving constructor,
to encourate moves for movable types

Signed-off-by: Sawyer Bergeron <sawyerbergeron@gmail.com>

szbergeron committed:
Fix docstr on alternate copying constructor for Box[]

Signed-off-by: Sawyer Bergeron <sawyerbergeron@gmail.com>

szbergeron committed:
Refmt

Signed-off-by: Sawyer Bergeron <sawyerbergeron@gmail.com>

szbergeron committed:
Doc dummy param

Signed-off-by: Sawyer Bergeron <sawyerbergeron@gmail.com>
